### PR TITLE
Feature Requested By my Friend @Kermit iykyk

### PR DESCRIPTION
**Pull request by rudyrdx**
# Pull Request Description

**Hello,**

My friend Kermit approached me with an issue where a JS script file was being stored inside the "Scripts" folder.

This pull request still saves the file as intended, but it does not save the JS code inside it.

There were a few changes I had to make while debugging. The `Invoke` method, instead of asynchronously reading a file using `StaticNodeJSService.InvokeFromFileAsync`, now reads a string using `StaticNodeJSService.InvokeFromStringAsync`.

While this may seem like a minor change code-wise, it is a significant feature that my friend wanted.

The method `public static async Task<T> InvokeNode<T>(BotData data, string scriptOrFile, object[] parameters, bool isScript = false)` has been slightly modified to handle this case in the `Method.cs` file, as well as in line 199 of the `ScriptBlockInstance.cs` file:

```csharp
writer.WriteLine($"var {resultName} = await InvokeNode<dynamic>(data, {escapedScript}, new object[] {{ {InputVariables} }}, true);");
```

### CODE REVIEW

File: RuriLib/Blocks/Interop/Methods.cs
Comment (openbullet) [inline RuriLib/Blocks/Interop/Methods.cs:53]: The whole idea behind using `InvokeFromFileAsync` is that, since this script can be called very frequently by multiple bots, it is better to cache it. According to the [documentation](https://github.com/JeringTech/Javascript.NodeJS), when using `InvokeFromFileAsync`, the file is cached using its absolute path on disk. When using `InvokeFromStringAsync`, we would need to provide a cache key, otherwise it can have an impact on performance.

I think in this case the cache key could be the MD5 of the script's string. Take a look at the repo's docs for a sample usage, trying to invoke from cache first using `TryInvokeFromCacheAsync`, and then on a cache miss using `InvokeFromStringAsync` by passing the cache key so it's cached.

File: RuriLib/Models/Blocks/Custom/ScriptBlockInstance.cs
Comment (openbullet) [inline RuriLib/Models/Blocks/Custom/ScriptBlockInstance.cs:178]: Please remove this, it's not very professional. I understand you did it for Kermit, but you can keep it in the PR's title / commit messages without littering the code 😅

File: RuriLib/Models/Blocks/Custom/ScriptBlockInstance.cs
Comment (openbullet) [inline RuriLib/Models/Blocks/Custom/ScriptBlockInstance.cs:186]: This is unprofessional, please remove. Also, writing something to a file is not needed anymore. We will keep the duality in the `InvokeNode<T>` method for backwards compatibility, in case of configs that were created with the old versions of RuriLib, but it's not required to write anything to a script file anymore, so it can all be removed. Only the `escapedScript` should be kept and directly injected into the code.

**Review by openbullet — CHANGES_REQUESTED** [review]
Hello,

first of all, thanks for the PR. I left a few comments on what should be changed before this can be merged.

I would also like to point out that the same thing could also be done for the other two supported engines (Jint and IronPython), so if you don't intend to do it, please open an issue and reference this PR so I can remember to do it one day 😉 

Let me know if you agree with the changes. Thanks!

**Comment by rudyrdx** [comment]
Ok good Sir!

File: OpenBullet2.Web/package-lock.json
Comment (openbullet) [inline OpenBullet2.Web/package-lock.json:1]: Why has this file been committed?

File: OpenBullet2.Web/package.json
Comment (openbullet) [inline OpenBullet2.Web/package.json:1]: Why has this file being committed?

File: RuriLib/Blocks/Interop/Methods.cs
Comment (openbullet) [inline RuriLib/Blocks/Interop/Methods.cs:44]: This is a breaking change to the signature of the function, old configs that have been compiled with an old build will not work anymore. A valid signature should be

```cs
public static async Task<T> InvokeNode<T>(BotData data, string scriptOrFile, object[] parameters, bool isScript = false, string scriptHash = null)
```

File: RuriLib/Blocks/Interop/Methods.cs
Comment (openbullet) [inline RuriLib/Blocks/Interop/Methods.cs:56]: This has to be changed anyways, according to the comment above, but I suggest that you use named tuples in the future instead of standard tuples. `Item1` and `Item2` are not good names.

File: RuriLib/Blocks/Interop/Methods.cs
Comment (openbullet) [inline RuriLib/Blocks/Interop/Methods.cs:64]: If we reached this line, `script.Item2` should contain an MD5 hash. Why is it being treated like a file path?

File: RuriLib/Blocks/Interop/Methods.cs
Comment (openbullet) [inline RuriLib/Blocks/Interop/Methods.cs:62]: Please separate methods with a blank line, according to C# coding style guidelines. Also, accept two individual parameters for `script` and `scriptHash` instead of a tuple.

File: RuriLib/Blocks/Interop/Methods.cs
Comment (openbullet) [inline RuriLib/Blocks/Interop/Methods.cs:86]: Please separate methods with a blank line, according to C# coding style guidelines. Also why does this method exist? If it's only called once and only does a single thing, there's no need to extract to a method at all.

File: RuriLib/Blocks/Interop/Methods.cs
Comment (openbullet) [inline RuriLib/Blocks/Interop/Methods.cs:94]: Please separate methods with a blank line, according to C# coding style guidelines.

File: RuriLib/Models/Blocks/Custom/ScriptBlockInstance.cs
Comment (openbullet) [inline RuriLib/Models/Blocks/Custom/ScriptBlockInstance.cs:192]: Why is `scriptPath` being used as a hash here? `scriptHash` is enough. There is no need to use a file anymore, the file-based invocation will only stay in the `InvokeNode` method for backwards compatibility, so `scriptPath` can be removed. Also, since it's just a hex string, it does not need to be escaped in this case, and can be passed to the writer like

```cs
writer.WriteLine($"var {resultName} = await InvokeNode<dynamic>(data, {escapedScript}, new object[] {{ {InputVariables} }}, true, \"{scriptHash}\");");
```

**Review by openbullet — CHANGES_REQUESTED** [review]
I'm sorry for writing so many comments, but I cannot approve this PR as it is now. If you don't have time (or do not want) to work on this, I can continue from where you left off by myself on another branch, but I'm not sure you would be marked as a contributor that way. Let me know please! Thanks 😄

**Comment by rudyrdx** [comment]
Hi, I have done all the suggested changes, this is my first time contributing to any source. Thanks for bearing me up till now.
If you'd want me to make the changes for Jint & IronPython, i can give them a try.

File: RuriLib/Blocks/Interop/Methods.cs
Comment (openbullet) [inline RuriLib/Blocks/Interop/Methods.cs:48]: If `isScript` is `true` but `scriptHash` is null or empty, it will try to (wrongly) invoke from file. This should never happen in our case, but it can happen in general depending on who calls the method, so we need to handle it. It can be rewritten as
```cs
if (isScript)
```
and `InvokeFromStringOrCache<T>` should expect a nullable string as the `scriptHash`, since the hash is not mandatory (even though we expect it to be there in our case for caching).

File: RuriLib/Blocks/Interop/Methods.cs
Comment (openbullet) [inline RuriLib/Blocks/Interop/Methods.cs:61]: `scriptHash` should be a nullable string, to match what said above. If it's null or empty, use `InvokeFromStringAsync` directly (using a null hash), otherwise try to invoke from cache like it is now.

**Comment by openbullet** [comment]
LGTM! Thanks for the contribution ;)

**Pull request merged**
